### PR TITLE
SONARJAVA-1313 Use LiveVariableAnalysis as purge for SymbolicExecution

### PR DESCRIPTION
**Pull request by mpaladin**

### CODE REVIEW

File: java-squid/src/main/java/org/sonar/java/collections/AVLTree.java [hunk 1]
```diff
@@ -113,11 +119,59 @@ private void forEach(AVLTree t, PMap.Consumer<K, V> action) {
...
+    @Override
+    public Map.Entry<K, V> next() {
+      if (!hasNext()) {
+        throw new NoSuchElementException();
+      }
+      if (stack.isEmpty()) {
+        descendLeft(current);
+      }
+      AVLTree node = stack.removeLast();
+      if (node.right() != EMPTY) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/collections/AVLTree.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)

File: java-squid/src/main/java/org/sonar/java/collections/AVLTree.java [hunk 2]
```diff
@@ -113,11 +119,59 @@ private void forEach(AVLTree t, PMap.Consumer<K, V> action) {
...
+      current = node;
+    }
+
+    @Override
+    public boolean hasNext() {
+      return current != null;
+    }
+
+    private void descendLeft(AVLTree node) {
+      if (node != EMPTY) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/collections/AVLTree.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)

File: java-squid/src/main/java/org/sonar/java/collections/AVLTree.java [hunk 3]
```diff
@@ -291,14 +345,22 @@ public boolean isEmpty() {
       return false;
     }
 
+    private class Hasher implements PMap.Consumer<Object, Object> {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/collections/AVLTree.java]: MAJOR Make this a "static" inner class. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2694)

File: java-squid/src/main/java/org/sonar/java/se/ProgramState.java [hunk 4]
```diff
@@ -172,13 +178,116 @@ ProgramState put(Symbol symbol, SymbolicValue value) {
...
+    return this;
+  }
+
+  private static boolean isDisposable(SymbolicValue symbolicValue, @Nullable Object constraint) {
+    return SymbolicValue.isDisposable(symbolicValue) && (constraint == null || !(constraint instanceof ObjectConstraint) || ((ObjectConstraint) constraint).isDisposable());
+  }
+
+  private boolean inStack(SymbolicValue symbolicValue) {
+    for (SymbolicValue value : stack) {
+      if (symbolicValue == value) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/se/ProgramState.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)

File: java-squid/src/main/java/org/sonar/java/se/ProgramState.java [hunk 5]
```diff
@@ -172,13 +178,116 @@ ProgramState put(Symbol symbol, SymbolicValue value) {
...
+
+  private static boolean inBinarySymbolicValue(SymbolicValue symbolicValue, SymbolicValue.BinarySymbolicValue value) {
+    Deque<SymbolicValue.BinarySymbolicValue> queue = new ArrayDeque<>();
+    queue.addLast(value);
+    while (!queue.isEmpty()) {
+      SymbolicValue.BinarySymbolicValue last = queue.removeLast();
+      for (SymbolicValue current : ImmutableList.of(last.leftOp, last.rightOp)) {
+        if (current instanceof SymbolicValue.BinarySymbolicValue) {
+          queue.addLast((SymbolicValue.BinarySymbolicValue) current);
+        } else if (current == symbolicValue) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/se/ProgramState.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)

File: java-squid/src/main/java/org/sonar/java/se/ProgramState.java [hunk 6]
```diff
@@ -68,10 +70,12 @@ public Pop(ProgramState programState, List<SymbolicValue> result) {
 
   private final Deque<SymbolicValue> stack;
   private final PMap<Symbol, SymbolicValue> values;
+  private final PMap<SymbolicValue, Integer> references;
   private final PMap<SymbolicValue, Object> constraints;
-  private ProgramState(PMap<Symbol, SymbolicValue> values, PMap<SymbolicValue, Object> constraints, PMap<ExplodedGraph.ProgramPoint, Integer> visitedPoints,
+  private ProgramState(PMap<Symbol, SymbolicValue> values, PMap<SymbolicValue, Integer> references, PMap<SymbolicValue, Object> constraints, PMap<ExplodedGraph.ProgramPoint, Integer> visitedPoints,
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/se/ProgramState.java]: MINOR Split this 197 characters long line (which is greater than 180 authorized). [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS00103)

**Comment by SonarTech** [comment]
SonarQube analysis reported 7 issues:
* MAJOR 6 major
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Change this comparison to use the equals method.](https://github.com/SonarSource/sonar-java/blob/d0f8dbad97d142f2174660d7b1f918146d2d9098/java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java#L250) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* MAJOR 4 major

Watch the comments in this conversation to review them.

File: java-squid/src/main/java/org/sonar/java/se/ProgramState.java [hunk 7]
```diff
@@ -172,13 +179,137 @@ ProgramState put(Symbol symbol, SymbolicValue value) {
...
     return this;
   }
 
+  private static boolean isDisposable(SymbolicValue symbolicValue, @Nullable Object constraint) {
+    return SymbolicValue.isDisposable(symbolicValue) && (constraint == null || !(constraint instanceof ObjectConstraint) || ((ObjectConstraint) constraint).isDisposable());
+  }
+
+  private static boolean inStack(Deque<SymbolicValue> stack, SymbolicValue symbolicValue) {
+    for (SymbolicValue value : stack) {
+      if (symbolicValue == value) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/se/ProgramState.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 5 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Change this comparison to use the equals method.](https://github.com/SonarSource/sonar-java/blob/69995d852ecc3b200af9f63d53a25142e9fd8b3e/java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java#L251) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* MAJOR 4 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* MAJOR 4 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* MAJOR 4 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 5 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Change this comparison to use the equals method.](https://github.com/SonarSource/sonar-java/blob/ac41c52ddf034aeb0fbb4a9e6dd21fe2e4c95e9f/java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java#L251) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* MAJOR 4 major

Watch the comments in this conversation to review them.

File: java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java [hunk 8]
```diff
@@ -202,9 +203,15 @@ public ProgramState apply(ProgramState input) {
     return startingStates;
   }
 
+  private void cleanUpProgramState(CFG.Block block) {
+    programState = ProgramState.cleanupDeadSymbols(programState, liveVariables.getOut(block));
```
Comment (benzonico) [inline java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java]: why a static method here ? for constancy I would make it not static.

File: java-squid/src/main/java/org/sonar/java/collections/AVLTree.java [hunk 9]
```diff
@@ -113,11 +119,53 @@ private void forEach(AVLTree t, PMap.Consumer<K, V> action) {
...
+      descendLeft(node);
+    }
+
+    @Override
+    public boolean hasNext() {
+      return !stack.isEmpty();
+    }
+
+    private void descendLeft(AVLTree node) {
+      if (node != EMPTY) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/collections/AVLTree.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)

File: java-squid/src/main/java/org/sonar/java/collections/AVLTree.java [hunk 10]
```diff
@@ -113,11 +119,53 @@ private void forEach(AVLTree t, PMap.Consumer<K, V> action) {
...
+      }
+    }
+
+    @Override
+    public Map.Entry<K, V> next() {
+      if (!hasNext()) {
+        throw new NoSuchElementException();
+      }
+      AVLTree node = stack.removeLast();
+      if (node.right() != EMPTY) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/collections/AVLTree.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 3 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Change this comparison to use the equals method.](https://github.com/SonarSource/sonar-java/blob/75476ed8890037e24903fd63cdc0a50d5827c08c/java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java#L261) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 3 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Change this comparison to use the equals method.](https://github.com/SonarSource/sonar-java/blob/e5779fec6b0690b5b623cfb612f08db091db210c/java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java#L262) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 3 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Change this comparison to use the equals method.](https://github.com/SonarSource/sonar-java/blob/af043efb01fbd18e379c80d1617fc887100a694e/java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java#L262) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)

**Pull request merged**